[Bug]: error when trying to serialize headers with LongString

### What happened?

Apache Beam version: 2.55.0
Java version: 17

When attempting to read messages from RabbitMQ with List type headers (streaming dataflow), a NotSerializableException error is being returned.

Example:
![image](https://github.com/apache/beam/assets/32180411/f7acc148-fd20-4e3b-b630-57889a3ce164)

If the headers are String, Number, or Boolean, it works normally. I have tried using a CustomCoder with `.setCoder(new CustomCoder())` to try to avoid the headers that are not important in this case, but it appears to throw the error before reaching this coder.

Code:
```
PCollection<RabbitMqMessage> rabbitMessagesCollection = pipeline
            .apply("Read RabbitMQ Messages", RabbitMqIO.read()
                    .withUri(options.getRabbitURI())
                    .withQueue(options.getInputQueue()))
```

Stacktrace:
```
Error message from worker: java.lang.IllegalArgumentException: Unable to encode element 'ValueWithRecordId{id=[], value=org.apache.beam.sdk.io.rabbitmq.RabbitMqMessage@f9a0d014}' with coder 'ValueWithRecordId$ValueWithRecordIdCoder(SerializableCoder(org.apache.beam.sdk.io.rabbitmq.RabbitMqMessage))'.
        org.apache.beam.sdk.coders.Coder.getEncodedElementByteSize(Coder.java:300)
        org.apache.beam.sdk.coders.Coder.registerByteSizeObserver(Coder.java:291)
        org.apache.beam.sdk.util.WindowedValue$FullWindowedValueCoder.registerByteSizeObserver(WindowedValue.java:642)
        org.apache.beam.sdk.util.WindowedValue$FullWindowedValueCoder.registerByteSizeObserver(WindowedValue.java:558)
        org.apache.beam.runners.dataflow.worker.IntrinsicMapTaskExecutorFactory$ElementByteSizeObservableCoder.registerByteSizeObserver(IntrinsicMapTaskExecutorFactory.java:384)
        org.apache.beam.runners.dataflow.worker.util.common.worker.OutputObjectAndByteCounter.update(OutputObjectAndByteCounter.java:128)
        org.apache.beam.runners.dataflow.worker.DataflowOutputCounter.update(DataflowOutputCounter.java:67)
        org.apache.beam.runners.dataflow.worker.util.common.worker.OutputReceiver.process(OutputReceiver.java:48)
        org.apache.beam.runners.dataflow.worker.util.common.worker.ReadOperation.runReadLoop(ReadOperation.java:218)
        org.apache.beam.runners.dataflow.worker.util.common.worker.ReadOperation.start(ReadOperation.java:169)
        org.apache.beam.runners.dataflow.worker.util.common.worker.MapTaskExecutor.execute(MapTaskExecutor.java:83)
        org.apache.beam.runners.dataflow.worker.StreamingDataflowWorker.process(StreamingDataflowWorker.java:1263)
        org.apache.beam.runners.dataflow.worker.StreamingDataflowWorker.lambda$scheduleWorkItem$11(StreamingDataflowWorker.java:999)
        org.apache.beam.runners.dataflow.worker.streaming.Work.run(Work.java:81)
        org.apache.beam.runners.dataflow.worker.util.BoundedQueueExecutor.lambda$executeLockHeld$0(BoundedQueueExecutor.java:183)
        java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
        java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
        java.base/java.lang.Thread.run(Thread.java:833)
Caused by: java.io.NotSerializableException: com.rabbitmq.client.impl.LongStringHelper$ByteArrayLongString
        java.base/java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1187)
        java.base/java.io.ObjectOutputStream.writeObject(ObjectOutputStream.java:350)
        java.base/java.util.ArrayList.writeObject(ArrayList.java:866)
        java.base/jdk.internal.reflect.GeneratedMethodAccessor62.invoke(Unknown Source)
        java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        java.base/java.lang.reflect.Method.invoke(Method.java:568)
        java.base/java.io.ObjectStreamClass.invokeWriteObject(ObjectStreamClass.java:1070)
        java.base/java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1516)
        java.base/java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1438)
        java.base/java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1181)
        java.base/java.io.ObjectOutputStream.writeObject(ObjectOutputStream.java:350)
        java.base/java.util.HashMap.internalWriteEntries(HashMap.java:1944)
        java.base/java.util.HashMap.writeObject(HashMap.java:1497)
        java.base/jdk.internal.reflect.GeneratedMethodAccessor61.invoke(Unknown Source)
        java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        java.base/java.lang.reflect.Method.invoke(Method.java:568)
        java.base/java.io.ObjectStreamClass.invokeWriteObject(ObjectStreamClass.java:1070)
        java.base/java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1516)
        java.base/java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1438)
        java.base/java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1181)
        java.base/java.io.ObjectOutputStream.defaultWriteFields(ObjectOutputStream.java:1572)
        java.base/java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1529)
        java.base/java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1438)
        java.base/java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1181)
        java.base/java.io.ObjectOutputStream.defaultWriteFields(ObjectOutputStream.java:1572)
        java.base/java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1529)
        java.base/java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1438)
        java.base/java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1181)
        java.base/java.io.ObjectOutputStream.writeObject(ObjectOutputStream.java:350)
        org.apache.beam.sdk.coders.SerializableCoder.encode(SerializableCoder.java:192)
        org.apache.beam.sdk.coders.SerializableCoder.encode(SerializableCoder.java:57)
        org.apache.beam.sdk.values.ValueWithRecordId$ValueWithRecordIdCoder.encode(ValueWithRecordId.java:106)
        org.apache.beam.sdk.values.ValueWithRecordId$ValueWithRecordIdCoder.encode(ValueWithRecordId.java:100)
        org.apache.beam.sdk.values.ValueWithRecordId$ValueWithRecordIdCoder.encode(ValueWithRecordId.java:82)
        org.apache.beam.sdk.coders.Coder.getEncodedElementByteSize(Coder.java:297)
        org.apache.beam.sdk.coders.Coder.registerByteSizeObserver(Coder.java:291)
        org.apache.beam.sdk.util.WindowedValue$FullWindowedValueCoder.registerByteSizeObserver(WindowedValue.java:642)
        org.apache.beam.sdk.util.WindowedValue$FullWindowedValueCoder.registerByteSizeObserver(WindowedValue.java:558)
        org.apache.beam.runners.dataflow.worker.IntrinsicMapTaskExecutorFactory$ElementByteSizeObservableCoder.registerByteSizeObserver(IntrinsicMapTaskExecutorFactory.java:384)
        org.apache.beam.runners.dataflow.worker.util.common.worker.OutputObjectAndByteCounter.update(OutputObjectAndByteCounter.java:128)
        org.apache.beam.runners.dataflow.worker.DataflowOutputCounter.update(DataflowOutputCounter.java:67)
        org.apache.beam.runners.dataflow.worker.util.common.worker.OutputReceiver.process(OutputReceiver.java:48)
        org.apache.beam.runners.dataflow.worker.util.common.worker.ReadOperation.runReadLoop(ReadOperation.java:218)
        org.apache.beam.runners.dataflow.worker.util.common.worker.ReadOperation.start(ReadOperation.java:169)
        org.apache.beam.runners.dataflow.worker.util.common.worker.MapTaskExecutor.execute(MapTaskExecutor.java:83)
        org.apache.beam.runners.dataflow.worker.StreamingDataflowWorker.process(StreamingDataflowWorker.java:1263)
        org.apache.beam.runners.dataflow.worker.StreamingDataflowWorker.lambda$scheduleWorkItem$11(StreamingDataflowWorker.java:999)
        org.apache.beam.runners.dataflow.worker.streaming.Work.run(Work.java:81)
        org.apache.beam.runners.dataflow.worker.util.BoundedQueueExecutor.lambda$executeLockHeld$0(BoundedQueueExecutor.java:183)
        java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
        java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
        java.base/java.lang.Thread.run(Thread.java:833)
```


### Issue Priority

Priority: 2 (default / most bugs should be filed as P2)

### Issue Components

- [ ] Component: Python SDK
- [X] Component: Java SDK
- [ ] Component: Go SDK
- [ ] Component: Typescript SDK
- [ ] Component: IO connector
- [ ] Component: Beam YAML
- [ ] Component: Beam examples
- [ ] Component: Beam playground
- [ ] Component: Beam katas
- [ ] Component: Website
- [ ] Component: Spark Runner
- [ ] Component: Flink Runner
- [ ] Component: Samza Runner
- [ ] Component: Twister2 Runner
- [ ] Component: Hazelcast Jet Runner
- [X] Component: Google Cloud Dataflow Runner

--- a/sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/RabbitMqMessage.java
+++ b/sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/RabbitMqMessage.java
@@ -27,8 +27,11 @@ import java.nio.charset.StandardCharsets;
 import java.util.Arrays;
 import java.util.Date;
 import java.util.HashMap;
+import java.util.List;
 import java.util.Map;
 import java.util.Objects;
+import java.util.stream.Collectors;
+import org.apache.beam.vendor.guava.v32_1_2_jre.com.google.common.annotations.VisibleForTesting;
 import org.checkerframework.checker.nullness.qual.Nullable;
 
 /**
@@ -73,33 +76,43 @@ public class RabbitMqMessage implements Serializable {
         envelope, nextProperties, processed.getBody(), processed.getMessageCount());
   }
 
-  private static Map<String, Object> serializableHeaders(Map<String, Object> headers) {
+  @VisibleForTesting
+  static Map<String, Object> serializableHeaders(Map<String, Object> headers) {
     Map<String, Object> returned = new HashMap<>();
     if (headers != null) {
       for (Map.Entry<String, Object> h : headers.entrySet()) {
         Object value = h.getValue();
-        if (!(value instanceof Serializable)) {
-          try {
-            if (value instanceof LongString) {
-              LongString longString = (LongString) value;
-              byte[] bytes = longString.getBytes();
-              String s = new String(bytes, StandardCharsets.UTF_8);
-              value = s;
-            } else {
-              throw new RuntimeException(String.format("no transformation defined for %s", value));
-            }
-          } catch (Throwable t) {
-            throw new UnsupportedOperationException(
-                String.format(
-                    "can't make unserializable value %s a serializable value (which is mandatory for Apache Beam dataflow implementation)",
-                    value),
-                t);
-          }
+        if (value instanceof List<?>) {
+          // Transformation for List type headers
+          value =
+              ((List<?>) value)
+                  .stream().map(RabbitMqMessage::getTransformedValue).collect(Collectors.toList());
+        } else if (!(value instanceof Serializable)) {
+          value = getTransformedValue(value);
         }
         returned.put(h.getKey(), value);
       }
     }
     return returned;
+  }
+
+  private static Object getTransformedValue(Object value) {
+    try {
+      if (value instanceof LongString) {
+        LongString longString = (LongString) value;
+        byte[] bytes = longString.getBytes();
+        value = new String(bytes, StandardCharsets.UTF_8);
+      } else {
+        throw new RuntimeException(String.format("No transformation defined for %s", value));
+      }
+    } catch (Throwable t) {
+      throw new UnsupportedOperationException(
+          String.format(
+              "Can't make unserializable value %s a serializable value (which is mandatory for Apache Beam dataflow implementation)",
+              value),
+          t);
+    }
+    return value;
   }
 
   private final @Nullable String routingKey;

--- a/sdks/java/io/rabbitmq/src/test/java/org/apache/beam/sdk/io/rabbitmq/RabbitMqMessageTest.java
+++ b/sdks/java/io/rabbitmq/src/test/java/org/apache/beam/sdk/io/rabbitmq/RabbitMqMessageTest.java
@@ -1,0 +1,76 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.io.rabbitmq;
+
+import static org.junit.Assert.assertEquals;
+
+import com.rabbitmq.client.LongString;
+import com.rabbitmq.client.impl.LongStringHelper;
+import java.io.Serializable;
+import java.nio.charset.StandardCharsets;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import org.apache.beam.vendor.guava.v32_1_2_jre.com.google.common.collect.Lists;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+/** Test of {@link RabbitMqMessage}. */
+@RunWith(JUnit4.class)
+public class RabbitMqMessageTest implements Serializable {
+
+  @Test(expected = UnsupportedOperationException.class)
+  public void testSerializableHeadersThrowsIfValueIsNotSerializable() {
+    Map<String, Object> rawHeaders = new HashMap<>();
+    Object notSerializableObject = Optional.of(new Object());
+    rawHeaders.put("key1", notSerializableObject);
+    RabbitMqMessage.serializableHeaders(rawHeaders);
+  }
+
+  @Test
+  public void testSerializableHeadersWithLongStringValues() {
+    Map<String, Object> rawHeaders = new HashMap<>();
+    String key1 = "key1", key2 = "key2", value1 = "value1", value2 = "value2";
+    rawHeaders.put(key1, LongStringHelper.asLongString(value1));
+    rawHeaders.put(key2, LongStringHelper.asLongString(value2.getBytes(StandardCharsets.UTF_8)));
+
+    Map<String, Object> serializedHeaders = RabbitMqMessage.serializableHeaders(rawHeaders);
+
+    assertEquals(value1, serializedHeaders.get(key1));
+    assertEquals(value2, serializedHeaders.get(key2));
+  }
+
+  @Test
+  public void testSerializableHeadersWithListValue() {
+    Map<String, Object> rawHeaders = new HashMap<>();
+    List<String> expectedSerializedList = Lists.newArrayList("value1", "value2");
+    List<LongString> rawList =
+        expectedSerializedList.stream()
+            .map(LongStringHelper::asLongString)
+            .collect(Collectors.toList());
+    String key1 = "key1";
+    rawHeaders.put(key1, rawList);
+
+    Map<String, Object> serializedHeaders = RabbitMqMessage.serializableHeaders(rawHeaders);
+
+    assertEquals(expectedSerializedList, serializedHeaders.get(key1));
+  }
+}

